### PR TITLE
Replace (val / 2) with (val * 0.5) in Jit

### PR DESCRIPTION
**Pull request by EgorBo**
`vmulss/vmulsd` has better both latency and throughput than `vdivss/vdivsd` at least for the hardware I have. e.g. on my MacBook's Haswell:
```
vdivss (Latency: 10-20,  R.Throughput: 7-14)
vmulss (Latency:     5,  R.Throughput:  0.5)
```

So if a divisor is a constant power of two we can optimize it, e.g.:
```
a = b / 2;
becomes:
a = b * 0.5;
```
See https://godbolt.org/z/rz9h4E (clang, gcc, msvc, x86, AMD64, AArch64 - everywhere this optimization is applied. Btw, LLVM also helps Mono to optimize this case for C#)

I wrote a small benchmark:
```csharp
[Benchmark]
[Arguments(MathF.PI)]
public float Div_old(float value)
{
    for (int i = 0; i < 10; i++)
        value = value / 2f;
    return value;
}

[Benchmark(Baseline = true)]
[Arguments(MathF.PI)]
public float Div_new(float value)
{
    for (int i = 0; i < 10; i++)
        value = value * 0.5f;
    return value;
}
```
and the results are (Haswell):
```
BenchmarkDotNet=v0.11.5, OS=macOS Mojave 10.14 (18A391) [Darwin 18.0.0]
Intel Core i7-4980HQ CPU 2.80GHz (Haswell), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=3.0.100-preview5-011568
  [Host]   : .NET Core 3.0.0-preview5-27626-15 (CoreCLR 4.6.27622.75, CoreFX 4.700.19.22408), 64bit RyuJIT
  ShortRun : .NET Core 3.0.0-preview5-27626-15 (CoreCLR 4.6.27622.75, CoreFX 4.700.19.22408), 64bit RyuJIT

Job=ShortRun  IterationCount=3  LaunchCount=1
WarmupCount=3

|  Method |     value |      Mean |     Error |    StdDev | Ratio | RatioSD |
|-------- |---------- |----------:|----------:|----------:|------:|--------:|
| Div_old | 3.1415927 | 16.848 ns | 1.6733 ns | 0.0917 ns |  3.14 |    0.02 |
| Div_new | 3.1415927 |  5.359 ns | 0.5572 ns | 0.0305 ns |  1.00 |    0.00 |

```
**UPD** the difference is only 20-30% on Coffee Lake (i7 8700K). Didn't check ARM for it.

PS: Also I noticed that `x / 1` is not optimized to just `x` see [sharplab.io](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABABgAJiBGAbgFgAoYgZkoCZyBhRgb0fIGVWAMwA2EbBnIBBABRiJUhAEp+gvg0FbKAdnIJyAenK01AgL5nyVluQCWAOykAhWY6WrN6q1uJ6DxqZeFozmQA===)
(I found [this](https://github.com/dotnet/coreclr/blob/master/src/jit/morph.cpp#L14223-L14229) in morph.cpp but it seems it's not used for floats)

/cc: @tannergooding

### CODE REVIEW

File: src/jit/morph.cpp [hunk 1]
```diff
@@ -11571,7 +11571,18 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac)
             break;
 
         case GT_DIV:
-
+            /* Replace "val / dcon" with "val * (1.0 / dcon)" if dcon is a power of two */
```
Comment (tannergooding) [inline src/jit/morph.cpp]: It would be good to give a brief explanation of why this only handles powers of two. Namely, powers of two within range are always exactly represented, so multiplication by the reciprocal is safe in this scenario.

File: src/jit/morph.cpp [hunk 2]
```diff
@@ -11571,7 +11571,18 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac)
...
+            if ((typ == TYP_FLOAT || typ == TYP_DOUBLE) && op2->IsCnsFltOrDbl())
+            {
+                double op2dVal = op2->gtDblCon.gtDconVal;
+                int64_t op2iVal = static_cast<int64_t>(op2dVal);
+                if (op2dVal == op2iVal && isPow2(op2iVal))
```
Comment (tannergooding) [inline src/jit/morph.cpp]: I'd rather see this do an appropriate `isPow2(float)` and `isPow2(double)` check (which would also allow multiplication by reciprocal for values like `0.5`).

For normal finite numbers, this is basically just checking that the mantissa is zero. For subnormal finite numbers (where the exponent is zero), this is basically checking that the mantissa is itself a power of 2.

Comment (tannergooding) [reply src/jit/morph.cpp]: Speaking of multiplication by the reciprocal of `0.5`. It looks like we don't do optimizations like converting `x * 2.0` to `x + x` or `x * 1.0` or `x * 0.0`

File: src/jit/morph.cpp (hunk 1 above)
Comment (mikedn) [inline src/jit/morph.cpp]: In general `//` style comments are preferred in the JIT code.

File: src/jit/morph.cpp [hunk 3]
```diff
@@ -11571,7 +11571,18 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac)
             break;
 
         case GT_DIV:
-
+            /* Replace "val / dcon" with "val * (1.0 / dcon)" if dcon is a power of two */
+            if ((typ == TYP_FLOAT || typ == TYP_DOUBLE) && op2->IsCnsFltOrDbl())
+            {
+                double op2dVal = op2->gtDblCon.gtDconVal;
```
Comment (mikedn) [inline src/jit/morph.cpp]: Use `AsDblCon()` instead of `gtDblCon`.

File: src/jit/morph.cpp (hunk 2 above)
Comment (mikedn) [reply src/jit/morph.cpp]: Properly checking for power of 2 is basically a must. As is, the code triggers undefined behavior for large values and that's no good in a compiler.

File: src/jit/morph.cpp [hunk 4]
```diff
@@ -11571,7 +11571,18 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac)
...
+                int64_t op2iVal = static_cast<int64_t>(op2dVal);
+                if (op2dVal == op2iVal && isPow2(op2iVal))
+                {
+                    oper = GT_MUL;
+                    tree->ChangeOper(oper);
+                    op2->gtDblCon.gtDconVal = 1.0 / op2dVal;
```
Comment (mikedn) [inline src/jit/morph.cpp]: This is a bit problematic if denormals are treated as zero for whatever reason (someone PInvoked code that changed the FP environment). Perhaps it would be better that together with a proper power of 2 check to also reject exponents 1023 (for double) and 127 (for float).

File: src/jit/morph.cpp (hunk 2 above)
Comment (EgorBo) [reply src/jit/morph.cpp]: I am not sure I did it correctly could you please take a look at my isMod2 implementations for float/double? (can I use frexp here?)

File: src/jit/utils.h [hunk 5]
```diff
@@ -42,6 +42,22 @@ inline bool isPow2(T i)
     return (i > 0 && ((i - 1) & i) == 0);
 }
 
+inline bool isPow2f(float v)
+{
+    uint32_t i = *(uint32_t*)&v;
+    uint8_t exponent = (uint8_t)(i >> 23);
```
Comment (tannergooding) [inline src/jit/utils.h]: You need to strip the sign bit

File: src/jit/utils.h [hunk 6]
```diff
@@ -42,6 +42,22 @@ inline bool isPow2(T i)
...
+    uint32_t mantissa =  i & 0x7fffff;
+    return mantissa == 0 && exponent != 0 && exponent != 127;
+}
+
+inline bool isPow2d(double v)
+{
+    uint64_t i = *(uint64_t*)&v;
+    uint32_t exponent = (uint32_t)(i >> 52);
+    uint64_t mantissa =  i & 0x8ffffffffffff;
```
Comment (tannergooding) [inline src/jit/utils.h]: This should be `0xFFFFFFFFFFFFF` (`i & ((1 << 52) - 1))`)

File: src/jit/utils.h [hunk 7]
```diff
@@ -42,6 +42,22 @@ inline bool isPow2(T i)
...
+    return mantissa == 0 && exponent != 0 && exponent != 1023;
```
Comment (tannergooding) [inline src/jit/utils.h]: I think this should just return `true` if it is a power of two, even for denormals. The actual optimization code should then filter out denormal inputs and denormal reciprocals separately.

File: src/jit/utils.h (hunk 5 above)
Comment (EgorBo) [reply src/jit/utils.h]: @tannergooding  isn't it stripped by the uint8_t cast here?

Comment (tannergooding) [reply src/jit/utils.h]: ah right. Just `double` needs fixing then.

File: src/jit/utils.h (hunk 7 above)
Comment (mikedn) [reply src/jit/utils.h]: Actually this is correct as it is, it's only the function name that is kind of unfortunate. It's not that useful to know that a FP value is a power of 2, it's not like we're going to use shifts, bittests and similar stuff to manipulate FP values. What we're after is basically a precise reciprocal so a `hasPreciseReciprocal` or something like that would be preferable.

Also, this should be moved to `FloatingPointUtils`.

File: src/jit/morph.cpp (hunk 2 above)
Comment (mikedn) [reply src/jit/morph.cpp]: I suppose you could avoid bit twiddling and use the C math library functions instead. Something like:
```C++
bool HasInverse(double d)
{
    int exp;
    return (frexp(d, &exp) != 0) && (exp != 0) && isnormal(1.0 / d);
}
```
Though last time I checked VC++ had rather ugly implementations (non-intrinsics, inefficient) of the above functions.

**Comment by mikedn** [comment]
FWIW I tried to do this or a similar FP optimization in the past but didn't bother with it because it wasn't very clear how useful it is, the framework doesn't have a lot of FP code. But now we have WPF in .NET Core - its PresentationCore.dll and PresentationFramework.dll contain a bit more than 100 hits. So it's quite useful, while many developers do this optimization manually there are still enough opportunities.

I haven't noticed any regressions. In theory this can block CSE because VN doesn't understand that `x * 0.5 == x / 2.0`. That can be fixed later if necessary, VN has other similar issues anyway.

File: src/jit/utils.cpp [hunk 8]
```diff
@@ -1954,6 +1954,13 @@ float FloatingPointUtils::round(float x)
     return _copysignf(flrTempVal, x);
 }
 
+bool FloatingPointUtils::isPow2(double x)
+{
+    int exp;
+    double f = frexp(x, &exp);
+    return abs(f) == 0.5 && x != 0 && isnormal(1.0 / x);
```
Comment (tannergooding) [inline src/jit/utils.cpp]: This is making a lot of assumptions about the crt implementation being correct/IEEE compliant and the floating-point state being in the correct format...

I much preferred the more explicit implementation.

**Comment by EgorBo** [comment]
@mikedn @tannergooding sorry for the delayed response. I've refactored it to `frexp` (I need to double check if it's 100% portable) based on your `HasInverse`
the `float / const` pattern can be found in game engines too, already found several places.

File: src/jit/utils.cpp (hunk 8 above)
Comment (EgorBo) [reply src/jit/utils.cpp]: I wonder if I can use `> 0 && <= 0.5` instead of `== 0.5`

**Comment by EgorBo** [comment]
Also, I did some testing locally, e.g. https://gist.github.com/EgorBo/866a49334291c1ac3b108eb9341681ae (similar for double and for non-power-of-two constants)

File: src/jit/utils.cpp (hunk 8 above)
Comment (tannergooding) [reply src/jit/utils.cpp]: I don't think that will be correct, since it would then cover non-powers of two.

The best behavior, IMO, would be to continue checking the raw bits (which is also notably what Clang is doing here) and not rely on the built-in CRT library.

Comment (EgorBo) [reply src/jit/utils.cpp]: @tannergooding do you have a link to llvm piece of code for it? 🙂 I tried to find it in llvm and gcc codebases but with no luck 🙁

Comment (tannergooding) [reply src/jit/utils.cpp]: The relevant bits are here:
* https://github.com/llvm-mirror/llvm/blob/master/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp#L1112
* https://github.com/llvm-mirror/llvm/blob/master/lib/IR/Constants.cpp#L227
* https://github.com/llvm-mirror/llvm/blob/master/lib/Support/APFloat.cpp#L3614

Looks like we already have the relevant 3rd party notice for LLVM: https://github.com/dotnet/coreclr/blob/master/THIRD-PARTY-NOTICES.TXT#L109

Comment (mikedn) [reply src/jit/utils.cpp]: > gcc codebases

Watch out with that, it's GPL code...

File: src/jit/morph.cpp [hunk 9]
```diff
@@ -11571,7 +11571,19 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac)
...
 
         case GT_DIV:
-
+            // Replace "val / dcon" with "val * (1.0 / dcon)" if dcon is a power of two.
+            // Powers of two within range are always exactly represented, 
+            // so multiplication by the reciprocal is safe in this scenario
+            if (op2->IsCnsFltOrDbl())
+            {
+                double divisor = op2->AsDblCon()->gtDconVal;
+                if ((typ == TYP_FLOAT || typ == TYP_DOUBLE) && FloatingPointUtils::isPow2(divisor))
```
Comment (mikedn) [inline src/jit/morph.cpp]: It looks like you broke the `float` case, I don't see how having a `double` only version of `isPow2` deals with the case of `float` denormal.

File: src/pal/src/cruntime/math.cpp [hunk 10]
```diff
@@ -309,6 +309,44 @@ PALIMPORT double __cdecl PAL_exp(double x)
     return ret;
 }
 
+/*++
+Function:
+    frexp
+
+See MSDN.
+--*/
+PALIMPORT double __cdecl PAL_frexp(double x, int* exp)
```
Comment (mikedn) [inline src/pal/src/cruntime/math.cpp]: Hmm, seems like using `frexp` might be more trouble than it's worth it.

File: src/jit/utils.cpp (hunk 8 above)
Comment (EgorBo) [reply src/jit/utils.cpp]: Ok, thanks! Rolled back to bit twiddling impl (it's not final - I am now investigating the corner cases (denormals, etc)).

File: src/jit/utils.cpp [hunk 11]
```diff
@@ -1954,6 +1954,42 @@ float FloatingPointUtils::round(float x)
...
+}
+
+bool FloatingPointUtils::isNormal(float x)
+{
+    int32_t bits = *(int32_t*)&x;
+    bits &= 0x7FFFFFFF;
+    return (bits < 0x7F800000) && (bits != 0) && ((bits & 0x7F800000) != 0);
+}
+
+bool FloatingPointUtils::isPow2(double x)
```
Comment (tannergooding) [inline src/jit/utils.cpp]: I liked mikedn's suggestion of `hasPreciseReciprocal` since we aren't doing proper `isPow2` check, it also is more similar to the `hasExactInverseFP` name that clang uses.

**Comment by tannergooding** [comment]
> Also: should I take care about Big Endian?

I think ensuring it is handled now is probably a good idea.

**Comment by tannergooding** [comment]
If you go to the AzDO page for the formatting job, you can pull down a patch file that will resolve the failures when applied to the git repo (`git apply format.patch`)

File: src/jit/utils.cpp [hunk 12]
```diff
@@ -1954,6 +1954,42 @@ float FloatingPointUtils::round(float x)
...
+
+bool FloatingPointUtils::hasPreciseReciprocal(double x)
+{
+    if (!isNormal(x))
+        return false;
+
+    uint64_t i = *(uint64_t*)&x;
+    uint64_t exponent = (i >> 52) && 0x7FFul;
+    uint64_t mantissa = i & 0xFFFFFFFFFFFFFul;
+    return mantissa == 0 && exponent != 0 && exponent != 1023;
```
Comment (tannergooding) [inline src/jit/utils.cpp]: nit: Might be worth a comment explaining how this works (I believe @CarolEidt, generally prefers function headers to be added as well).

Namely, it's worth calling out that this is checking for powers of 2 and that we don't handle denormals since those may not be well-defined on some platforms (such as if the user modified the floating-point environment via a P/Invoke) and that this includes returning false both when the input is denormal and when the reciprocal is denormal (the exponent is zero).

Comment (CarolEidt) [reply src/jit/utils.cpp]: > I believe @CarolEidt, generally prefers function headers to be added as well

It's actually in the JIT coding conventions here: https://github.com/dotnet/coreclr/blob/master/Documentation/coding-guidelines/clr-jit-coding-conventions.md#9.4

Comment (EgorBo) [reply src/jit/utils.cpp]: Oh, forgot about this PR :( Added comments to `hasPreciseReciprocal` let me know if you also need comments for trivial `isNormal`.

Comment (EgorBo) [reply src/jit/utils.cpp]: From the doc it's not clear for me should I write comments in .cpp or .h

File: tests/src/JIT/opt/InstructionCombining/DivToMul.cs [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+
+public class Program
+{
+    private static int resultCode = 100;
+
+    public static int Main(string[] args)
+    {
+        // Some corner cases
+        var floatValues = new List<float>(
+            new[] { 0, 1, MathF.PI, MathF.E, float.MinValue, float.MinValue, int.MaxValue, int.MinValue, float.NegativeInfinity, float.PositiveInfinity, float.NaN, (float) double.MaxValue, (float) double.MinValue });
```
Comment (danmosemsft) [inline tests/src/JIT/opt/InstructionCombining/DivToMul.cs]: Is negative zero relevant?

Comment (EgorBo) [reply tests/src/JIT/opt/InstructionCombining/DivToMul.cs]: I've just added more cases. These values then passed to the tests additionally with `-`. So the tests actually have `0.0 / c` and `-0.0 / c` cases (as well as `x / 0.0` and `x / -0.0`)

**Comment by EgorBo** [comment]
@tannergooding @mikedn @danmosemsft so does it look good? the CI failures seem unrelated

File: src/jit/utils.cpp [hunk 14]
```diff
@@ -1954,6 +1954,68 @@ float FloatingPointUtils::round(float x)
     return _copysignf(flrTempVal, x);
 }
 
+bool FloatingPointUtils::isNormal(double x)
+{
+    int64_t bits = *(int64_t*)&x;
```
Comment (mikedn) [inline src/jit/utils.cpp]: `reinterpret_cast` should be used instead of C style cast

File: src/jit/utils.cpp [hunk 15]
```diff
@@ -1954,6 +1954,68 @@ float FloatingPointUtils::round(float x)
...
+
+bool FloatingPointUtils::hasPreciseReciprocal(float x)
+{
+    if (!isNormal(x))
+    {
+        return false;
+    }
+
+    uint32_t i        = *(uint32_t*)&x;
+    uint8_t  exponent = (uint8_t)(i >> 23);
```
Comment (mikedn) [inline src/jit/utils.cpp]: Why the cast to `uint8_t`?

File: src/jit/morph.cpp [hunk 16]
```diff
@@ -11540,7 +11540,20 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac)
...
 
         case GT_DIV:
-
+            // Replace "val / dcon" with "val * (1.0 / dcon)" if dcon is a power of two.
+            // Powers of two within range are always exactly represented,
+            // so multiplication by the reciprocal is safe in this scenario
+            if (op2->IsCnsFltOrDbl())
+            {
+                double divisor = op2->AsDblCon()->gtDconVal;
+                if ((typ == TYP_DOUBLE && FloatingPointUtils::hasPreciseReciprocal(divisor)) ||
```
Comment (mikedn) [inline src/jit/morph.cpp]: In general JIT codebase uses extra paranthesis for clarity: `(typ == TYP_DOUBLE) && FloatingPointUtils::hasPreciseReciprocal(divisor)`

There are similar cases below in the utilities functions.

File: tests/src/JIT/opt/InstructionCombining/DivToMul.cs [hunk 17]
```diff
@@ -0,0 +1,187 @@
...
+                float.NegativeInfinity, double.NegativeInfinity,
+                float.PositiveInfinity, double.PositiveInfinity,
+                float.NaN, double.NaN
+            });
+
+
+        // Also, add some random values
+        var random = new Random();
+        for (int i = 0; i < 100000; i++)
+            testValues.Add(random.NextDouble());
```
Comment (mikedn) [inline tests/src/JIT/opt/InstructionCombining/DivToMul.cs]: Let's hope that the code is actually correct and the test doesn't start failing randomly...

File: src/jit/utils.cpp (hunk 15 above)
Comment (EgorBo) [reply src/jit/utils.cpp]: @mikedn oh, it was a not really obvious way to drop the sign bit. Will replace the cast with `& 0xFF`

File: src/jit/utils.cpp (hunk 14 above)
Comment (EgorBo) [reply src/jit/utils.cpp]: done. replaced with `reinterpret_cast<uint64_t &>(x)`

File: src/jit/utils.cpp (hunk 15 above)
Comment (mikedn) [reply src/jit/utils.cpp]: > oh, it was a not really obvious way to drop the sign bit.

Exactly :smile:. Not only that it isn't obvious but some analyzers may complain about such truncation (I'm pretty sure VC++'s `/RTCc` would)

File: tests/src/JIT/opt/InstructionCombining/DivToMul.cs (hunk 17 above)
Comment (CarolEidt) [reply tests/src/JIT/opt/InstructionCombining/DivToMul.cs]: IMO it is a really bad thing to have tests in the CI system that use `Random` without an explicit seed. My preference is to:
- provide known values to cover the variety of cases (as you've done above)
- use Random with an explicit seed to add additional coverage

At a *very* minimum (and this is *not* my preference), a pseudo-random seed can be obtained and both provided to the constructor as well as printed with a failure message when the test doesn't pass.
cc @dotnet/jit-contrib for additional opinions on this.

Comment (EgorBo) [reply tests/src/JIT/opt/InstructionCombining/DivToMul.cs]: Makes sense! Added a fixed seed. Btw, it looks like there are other places in other tests where Random is used with it.

Comment (echesakovMSFT) [reply tests/src/JIT/opt/InstructionCombining/DivToMul.cs]: I would suggest to use `TestLibrary.Generator` instead of explicitly creating `Random()` https://github.com/dotnet/coreclr/blob/master/tests/src/Common/CoreCLRTestLibrary/Generator.cs

We've been using `TestLibrary.Generator` in other tests in coreclr (e.g. HardwareIntrinsics)

Comment (EgorBo) [reply tests/src/JIT/opt/InstructionCombining/DivToMul.cs]: Replaced with `TestLibrary.Generator.GetDouble()`

Comment (mikedn) [reply tests/src/JIT/opt/InstructionCombining/DivToMul.cs]: IMO using random numbers in compiler test is pointless. Typically you have a few interesting corner cases and the rest. Where the rest is something like billions of cases. Unless you take special care to skew the random number distribution in a rather special manner you're wasting your time.

This is no different. There are almost no corner cases for this optimization.

Comment (EgorBo) [reply tests/src/JIT/opt/InstructionCombining/DivToMul.cs]: @mikedn In general I agree with you, I used these tests in TDD-style approach (Test First) and then got a lot of failures during implementation. I can delete them (at least the Random part)

Comment (CarolEidt) [reply tests/src/JIT/opt/InstructionCombining/DivToMul.cs]: Deleting them seems reasonable to me.

**Comment by EgorBo** [comment]
A small Roslyn-based script to find places where this optimization can be applied (found some in various math/graphics related C# repositories): https://gist.github.com/EgorBo/74b034fe1936c43fcd0b42934322557c
Also, there are places in CoreFX where `* 0.5` is used (I guess authors knew it's better 😄) e.g.: https://github.com/dotnet/corefx/blob/master/src/System.Runtime.Numerics/src/System/Numerics/Complex.cs#L418

File: tests/src/JIT/opt/InstructionCombining/DivToMul.cs (hunk 17 above)
Comment (EgorBo) [reply tests/src/JIT/opt/InstructionCombining/DivToMul.cs]: Deleted the Random part, or you suggest to delete them all?

**Comment by mikedn** [comment]
> A small Roslyn-based script to find places where this optimization can be applied (found some in various math/graphics related C# repositories): 

Erm, why do you need such a contraption? Perhaps you're not aware how run JIT diffs?

**Comment by EgorBo** [comment]
@mikedn I wanted to quickly find such places without even downloading repositories (via HttpClient) 🙂. Also I made a list of patterns that LLVM is able to optimize ([InstCombine](https://github.com/llvm-mirror/llvm/tree/master/lib/Transforms/InstCombine) transforms) and looking for them in those repos.

**Comment by mikedn** [comment]
> I wanted to quickly find such places without even downloading repositories

That's not going to find the interesting case, where expressions like `x / 2.0` appear as a result of other JIT optimizations (inlining, constant folding & propagation etc.). Cases where `x / 2.0` literally appears in the code aren't very interesting, developers can easily change that to `x * 0.5` if they wish so. Adding a zillion of such expression transformations to the JIT might not be the best thing to do, especially in the current JIT state (morph is really a nightmare).

Anyway, here's a x64 FX diff:
```
Summary:
(Lower is better)
Total bytes of diff: 0 (0.00% of base)
0 total files with size differences (0 improved, 0 regressed), 129 unchanged.
0 total methods with size differences (0 improved, 0 regressed), 146772 unchanged.
3 files had text diffs but not size diffs.
Microsoft.Diagnostics.Tracing.TraceEvent.dasm had 12 diffs
System.Private.CoreLib.dasm had 8 diffs
Microsoft.DotNet.Cli.Utils.dasm had 4 diffs
Completed analysis in 28.14s
```
As I mentioned in a previous post, it's a bit of a strange case because `divss` and `mulss` instructions have the same size so nothing shows up in terms of size differences. Text diff shows that there are only 6 diffs in the entire framework. Not surprising as there's not a lot of floating point code in corelib and corefx.

**Comment by EgorBo** [comment]
@mikedn yeah, but as you mentioned earlier there are cases in `dotnet/wpf` repo:
```
ReachFramework\AlphaFlattener\BrushProxy.cs:
        (-B + root) / A / 2

ReachFramework\AlphaFlattener\BrushProxy.cs:
        (-B - root) / A / 2

ReachFramework\Serialization\DrawingContextFlattener.cs:
        (rRadSquared + rDot) / 2

ReachFramework\Serialization\DrawingContextFlattener.cs:
        (xEnd - xStart) / 2

ReachFramework\Serialization\DrawingContextFlattener.cs:
        (yEnd - yStart) / 2

ReachFramework\Serialization\DrawingContextFlattener.cs:
        (xEnd + xStart) / 2

ReachFramework\Serialization\DrawingContextFlattener.cs:
        (yEnd + yStart) / 2

ReachFramework\Serialization\VisualSerializer.cs:
        - (dstwidth - width) / 2

ReachFramework\Serialization\VisualSerializer.cs:
        (dstwidth - width) / 2

ReachFramework\Serialization\VisualSerializer.cs:
        - (dstheight - height) / 2

ReachFramework\Serialization\VisualSerializer.cs:
        (dstheight - height) / 2

PresentationCore\MS\internal\Ink\StrokeRenderer.cs:
        Math.PI * 3.0 / 2.0

PresentationCore\MS\internal\Shaping\CompositeTypefaceMetrics.cs:
        (-UnderlineOffsetDefaultInEm) / 2

PresentationCore\System\Windows\Media\CompositionTarget.cs:
        Double.MinValue / 2.0

PresentationCore\System\Windows\Media\CompositionTarget.cs:
        Double.MinValue / 2.0

PresentationCore\System\Windows\Media\GlyphRun.cs:
        (firstStopAdvance + currentAdvance) / 2.0

PresentationFramework\MS\Internal\documents\DocumentGrid.cs:
        (centerWidth - ExtentWidth) / 2.0

PresentationFramework\MS\Internal\documents\DocumentGrid.cs:
        (ViewportHeight - ExtentHeight) / 2.0

PresentationFramework\MS\Internal\documents\DocumentGrid.cs:
        (ExtentWidth - _lastRowChangeExtentWidth) / 2.0

PresentationFramework\MS\Internal\documents\TextBoxView.cs:
        (width - lineWidth) / 2

PresentationFramework\MS\Internal\Ink\PenCursorManager.cs:
        (width - originalWidth) / 2

PresentationFramework\MS\Internal\Ink\PenCursorManager.cs:
        (height - originalHeight) / 2

PresentationFramework\System\Windows\Controls\Grid.cs:
        (minPower + maxPower) / 2.0

PresentationFramework\System\Windows\Documents\CaretElement.cs:
        double.MaxValue/2

PresentationFramework\System\Windows\Documents\CaretElement.cs:
        double.MaxValue/2

PresentationCore\System\Windows\Media\Animation\KeySpline.cs:
        (_parameter + top) / 2

PresentationCore\System\Windows\Media\Animation\KeySpline.cs:
        (_parameter + bottom) / 2

PresentationCore\System\Windows\Media\Animation\KeySpline.cs:
        (bottom + top) / 2
```

**Comment by mikedn** [comment]
Unfortunately running diffs on wpf assemblies is a bit more tricky at the moment so I haven't done it again. The ~100 hits estimation from my previous post on the matter likely still stands.

File: tests/src/JIT/opt/InstructionCombining/DivToMul.cs (hunk 17 above)
Comment (CarolEidt) [reply tests/src/JIT/opt/InstructionCombining/DivToMul.cs]: I only meant to delete the Random part - thanks.

File: src/jit/utils.cpp [hunk 18]
```diff
@@ -2038,6 +2038,68 @@ float FloatingPointUtils::round(float x)
...
 }
 
+bool FloatingPointUtils::isNormal(double x)
+{
+    int64_t bits = reinterpret_cast<int64_t&>(x);
+    bits &= 0x7FFFFFFFFFFFFFFF;
+    return (bits < 0x7FF0000000000000) && (bits != 0) && ((bits & 0x7FF0000000000000) != 0);
+}
+
+bool FloatingPointUtils::isNormal(float x)
```
Comment (sandreenko) [inline src/jit/utils.cpp]: I do not like this bit checks, but without C++ isnormal I do not see any better solution.

File: tests/src/JIT/opt/InstructionCombining/DivToMul.cs [hunk 19]
```diff
@@ -0,0 +1,187 @@
...
+        if (expectedi != actuali)
+        {
+            resultCode--;
+            Console.WriteLine($"AssertEquals: {expected} != {actual}");
+        }
+    }
+
+    [MethodImpl(MethodImplOptions.NoInlining)]
+    private static T ConstToVar<T>(T v) => v;
+}
```
Comment (sandreenko) [inline tests/src/JIT/opt/InstructionCombining/DivToMul.cs]: Please add a newline.

File: tests/src/JIT/opt/InstructionCombining/DivToMul.csproj [hunk 20]
```diff
@@ -0,0 +1,34 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (sandreenko) [inline tests/src/JIT/opt/InstructionCombining/DivToMul.csproj]: The project files were updated recently, could you please copy one of new csproj here instead of this one in an old format?

**Comment by briansull** [comment]
For float (32-bit) divides it would also  always be a win to perform a  reciprocal multiply operation using a (64-bit) multiply and conversion back to  float (32-bit)

**Comment by tannergooding** [comment]
> For float (32-bit) divides it would also always be a win to perform a reciprocal multiply operation using a (64-bit) multiply and conversion back to float (32-bit)

Why would you do that rather than just doing the 32-bit reciprocal multiplication? For cases where it is known to be equivalent, just keeping it as a single-precision float would be more efficient..

**Comment by briansull** [comment]
It was to cover the non power of two divide by constant case:

```
        public static float DivGen_new(float value)
        {
            for (int i = 0; i < 10; i++)
            {
                double valueD  = (double) value;
	        double resultD = valueD * (1.0 / 3.0);
	        value = (float) resultD;
            }
            return value;
        }
```

However it turns out the the convert instructions are pretty slow, so I believe that this transformation loses:

```
G_M2669_IG03:		;; bbWeight=10   
       F30F5AC0             cvtss2sd  xmm0, xmm0
       F20F590500000000     mulsd    xmm0, qword ptr [reloc @RWD00]
       F20F5AC0             cvtsd2ss  xmm0, xmm0
       FFC0                 inc      eax
       83F80A               cmp      eax, 10
       7CE9                 jl       SHORT G_M2669_IG03
```

**Comment by EgorBo** [comment]
@briansull nice try anyway 🙂

File: src/jit/morph.cpp [hunk 21]
```diff
@@ -11457,7 +11457,20 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac)
...
+            // so multiplication by the reciprocal is safe in this scenario
+            if (op2->IsCnsFltOrDbl())
+            {
+                double divisor = op2->AsDblCon()->gtDconVal;
+                if (((typ == TYP_DOUBLE) && FloatingPointUtils::hasPreciseReciprocal(divisor)) ||
+                    ((typ == TYP_FLOAT) && FloatingPointUtils::hasPreciseReciprocal(forceCastToFloat(divisor))))
+                {
+                    oper = GT_MUL;
+                    tree->ChangeOper(oper);
+                    op2->AsDblCon()->gtDconVal = 1.0 / divisor;
```
Comment (tannergooding) [inline src/jit/morph.cpp]: It might be worth noting that this is safe and doing the single operation in single precision isn't required.

The paper `Innocuous Double Rounding of Basic Arithmetic Operations` provides a proof that a single primitive operation done to at least twice the precision of the target format does not incur error due to double-rounding (hence for `float divisor` doing `(float)(1.0 / divisor)` is the same as `(1.0f / divisor)`).

This is not a safe thing to do across multiple operations (you must downcast back to float after each individual operation) nor is it safe if one of the inputs could not be exactly represented as a `float` (e.g. if you have `double divisor` and `(float)divisor != divisor`).

**Comment by sandreenko** [comment]
Thanks @EgorBo.
